src/darwin/Framework/CHIP/EndpointDescription/MTRAttributeDescription.mm

```diff
@@ -0,0 +1,195 @@
+/**
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import "MTRBaseDevice_Internal.h"
+#import "MTRError_Internal.h"
+#import "MTRLogging_Internal.h"
+#import "NSDataSpanConversion.h"
+#import <Matter/MTRAttributeDescription.h>
+
+#include <lib/core/CHIPError.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/SafeInt.h>
+
+using namespace chip;
+
+@interface MTRAttributeDescription ()
+// When initialized, exactly one of singleValue and listValue will be non-nil.
+@property (nonatomic, strong, readonly, nullable) NSData * singleValue;
+@property (nonatomic, strong, readonly, nullable) NSArray<NSData *> * listValue;
+@end
+
+@implementation MTRAttributeDescription
+
+@dynamic value;
+
+- (nullable instancetype)initAttributeWithID:(NSNumber *)attributeID initialValue:(NSDictionary<NSString *, id> *)value writable:(BOOL)writable error:(NSError * __autoreleasing *)error
+{
+    auto attrIDValue = attributeID.unsignedLongLongValue;
+    if (!CanCastTo<AttributeId>(attrIDValue)) {
+        MTR_LOG_ERROR("MTRAttributeDescription provided too-large attribute ID: 0x%llx", attrIDValue);
+        if (error) {
+            *error = [MTRError errorForCHIPErrorCode:CHIP_ERROR_INVALID_ARGUMENT];
+        }
+        return nil;
+    }
+
+    auto attrId = static_cast<AttributeId>(attrIDValue);
+    if (!IsValidAttributeId(attrId)) {
+        MTR_LOG_ERROR("MTRAttributeDescription provided invalid attribute ID: 0x%" PRIx32, attrId);
+        if (error) {
+            *error = [MTRError errorForCHIPErrorCode:CHIP_ERROR_INVALID_ARGUMENT];
+        }
+        return nil;
+    }
+
+    NSData * singleValue;
+    NSMutableArray<NSData *> * listValue;
+
+    id dataType = value[MTRTypeKey];
```

Comment:
It feels like it would make sense for the encoding to happen as needed, rather than in the initializer. The API makes it look like the dictionary representation is the "canonical" representation so it seems surprising that the getter has to decode the value again each time it's called.

Reply:
A few things:

1. I want to validate the input here.  Encoding it is the way to go.
2. What we need internally is the encoded form.
3. This lets me avoid writing deep-cloning code for the input dictionary... which I would end up doing by serializing/deserializing anyway.
4. In practice, I don't expect this getter to get called much....  I mean, I guess I could remove it and its property altogether and just have the internal getters for the data.  Would that be preferable?

Reply:
I am now storing the dictionary itself as well as the encoded form.